### PR TITLE
REM-347

### PR DESCRIPTION
**Pull request by imtithalhalalo**
Adding Checkbook page
ca-checkbook
Adding form and grid layout 5 fields (bank lookup, firstCheckNo, lastCheckNo, size, issueDate) all required for the form 
Editing and deleting records

### CODE REVIEW

**Comment by Ghinwa-Arab** [comment]
image

Size field minimum number is 0, here I'm able to enter -1 and save

**Comment by Ghinwa-Arab** [comment]
image

clear button resetting the whole form except size field

File: src/pages/ca-checkbook/index.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    })
+
+    return { ...response, _startAt: _startAt }
+  }
+
+  const {
+    query: { data },
+    labels: _labels,
+    paginationParameters,
+    refetch,
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/index.js:37]: add invalidate here and remove the one below

File: src/pages/ca-checkbook/index.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        valueGetter: ({ row }) => formatDateDefault(row?.issueDate)
+    }
+  ]
+
+  const add = () => {
+    openForm()
+  }
+
+  const edit = obj => {
+    console.log(obj.record)
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/index.js:88]: remove log

File: src/pages/ca-checkbook/index.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    openForm(obj?.recordId)
+  }
+
+  function openForm(recordId) {
+    stack({
+      Component: CheckbookForm,
+      props: {
+        labels: _labels,
+        recordId: recordId,
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/index.js:97]: enough to pass recordId,

File: src/pages/ca-checkbook/index.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        maxAccess: access
+      },
+      width: 500,
+      height: 500,
+      title: _labels.checkbook
+    })
+  }
+
+  const del = async obj => {
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/index.js:106]: add try catch

File: src/pages/ca-checkbook/index.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+          onEdit={edit}
+          onDelete={del}
+          isLoading={false}
+          pageSize={50}
+          paginationType='api'
+          paginationParameters={paginationParameters}
+          refetch={refetch}
+          maxAccess={access}
+        />
+      </Grow>{' '}
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/index.js:134]: remove { ' ' }

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+      bankAccountRef: null,
+      bankAccountName: null
+    },
+    maxAccess,
+    enableReinitialize: true,
+    validateOnChange: true,
+    validationSchema: yup.object({
+      bankAccountId: yup.string().required(' '),
+      size: yup.number().required(' '),
+      firstCheckNo: yup.string().required(' '),
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:45]: don't pass empty string 
example:
bankAccountId: yup.string().required(),

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+      lastCheckNo: yup.string().required(' '),
+      issueDate: yup.string().required(' ')
+    }),
+    onSubmit: async obj => {
+      const recordId = obj.recordId
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:50]: remove and use obj.recordId directly below

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+      const data = {
+        ...obj,
+        issueDate: formatDateToApi(obj.issueDate),
+        recordId: recordId
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:55]: recordId: recordId
serve no purpose

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      }
+
+      const response = await postRequest({
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:58]: add try catch whenever you request an API

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                maxAccess={maxAccess}
+                onChange={formik.handleChange}
+                onClear={() => formik.setFieldValue('lastCheckNo', '')}
+                error={formik.touched.lastCheckNo && Boolean(formik.errors.lastCheckNo)}
+              />
+            </Grid>
+            <Grid item xs={12}>
+                <CustomNumberField
+                  name='size'
+                  type='text'
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:148]: it's necessary to pass type='text'??

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+                  required
+                  label={labels.size}
+                  value={formik.values.size}
+                  maxAccess={maxAccess}
+                  maxLength={100}
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:153]: maxLength used for checking number of digits entered & not for checking the value if exceeds 100

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+                  onChange={e =>
+                    formik.setValues({
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:155]: replace by 
  onChange={formik.handleChange}

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+                      ...formik.values,
+                      size: e.target.value
+                    })
+                  }
+                  onClear={() =>
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:160]: replace
 onClear={() => formik.setFieldValue('size', '')}

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            <Grid item xs={12}>
+                <CustomDatePicker
+                  name='issueDate'
+                  label={labels.issueDate}
+                  value={formik.values?.issueDate}
+                  required
+                  onChange={formik.setFieldValue}
+                  onClear={() => formik.setFieldValue('issueDate', '')}
+                  error={formik.touched.issueDate && Boolean(formik.errors.issueDate)}
+                  helperText={formik.touched.issueDate && formik.errors.issueDate}
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:178]: remove helperText

File: src/pages/ca-checkbook/forms/CheckbookForm.js (hunk 10 above)
Comment (imtithalhalalo) [reply src/pages/ca-checkbook/forms/CheckbookForm.js:148]: no not needed

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+      size: yup.number().min(0).max(99).required(),
+      firstCheckNo: yup.string().required(),
+      lastCheckNo: yup.string().required(),
+      issueDate: yup.string().required()
+    }),
+    onSubmit: async obj => {
+      const data = {
+        ...obj,
+        issueDate: formatDateToApi(obj.issueDate),
+        recordId: obj.recordId
```
Comment (Ghinwa-Arab) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:53]: remove recordId: obj.recordId

File: src/pages/ca-checkbook/forms/CheckbookForm.js [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+      issueDate: null,
+      bankAccountRef: null,
+      bankAccountName: null
+    },
+    maxAccess,
+    enableReinitialize: true,
+    validateOnChange: true,
+    validationSchema: yup.object({
+      bankAccountId: yup.string().required(),
+      size: yup.number().min(0).max(99).required(),
```
Comment (FatimaSalami) [inline src/pages/ca-checkbook/forms/CheckbookForm.js:44]: when the size is greater than the limit the field is red with no message; i guess the user will not understand how to fix this

File: src/repositories/CashBankRepository.js [hunk 17]
```diff
@@ -76,5 +76,13 @@ export const CashBankRepository = {
...
     del: service + 'delADJ',
     post: service + 'postADJ'
+  },
+  CACheckbook: {
+    qry: service + 'qryCBK',
+    page: service + 'pageCBK',
+    get: service + 'getCBK',
+    set: service + 'setCBK',
+    del: service + 'delCBK',
+    post: service + 'postCBK'
```
Comment (FatimaSalami) [inline src/repositories/CashBankRepository.js:86]: do we need postCBK?

Comment (imtithalhalalo) [reply src/repositories/CashBankRepository.js:86]: no

**Pull request merged**